taraxa/state/dpos/precompiled/dpos_contract.go

```diff
@@ -907,26 +917,24 @@ func (self *Contract) registerValidatorWithoutChecks(ctx vm.CallFrame, block typ
 		return ErrBrokenState
 	}
 
-	// ctx.Account == contract address. Substract tokens that were sent to the contract as delegation
-	ctx.Account.SubBalance(ctx.Value)
-
 	state = new(State)
 	state.RewardsPer1Stake = bigutil.Big0
 
 	// Creates validator related objects in storage
-	self.validators.CreateValidator(owner_address, &args.Validator, block, ctx.Value, args.Commission, args.Description, args.Endpoint)
+	validator := self.validators.CreateValidator(owner_address, &args.Validator, block, args.Commission, args.Description, args.Endpoint)
 	state.Count++
 
 	if ctx.Value.Cmp(bigutil.Big0) == 1 {
-		new_vote_count := voteCount(ctx.Value, self.cfg.EligibilityBalanceThreshold, self.cfg.VoteEligibilityBalanceStep)
-		if new_vote_count > 0 {
-			self.eligible_vote_count = add64p(self.eligible_vote_count, new_vote_count)
+		amount_err := self.delegate_check_amount(ctx, validator)
+		if amount_err != nil {
+			return amount_err
```

Comment:
is this ok? are we should validator creation will be reverted

Reply:
If transaction is reverted all made changes should be reverted automatically, I think it is also applies for this precompiled contracts
Also we aren't reaching `state_put` method, so all of this changes won't be saved in db

Reply:
but you are calling create validator that is writing to the storage

Reply:
```
	if precompiled := self.precompiles.Get(code_owner.Address()); precompiled != nil {
		if gas_required := precompiled.RequiredGas(frame, self); gas_required <= gas_left {
			gas_left -= gas_required
			ret, err = precompiled.Run(frame, self)
		} else {
			err = ErrOutOfGas
		}
	} else if code := code_owner.GetCode(); len(code) != 0 {
		contract := NewContract(frame, CodeAndHash{code, code_owner.GetCodeHash()})
		ret, err = self.run(&contract, read_only)
		gas_left = contract.Gas
	}
	if err != nil {
		**self.state.RevertToSnapshot(snapshot)**
		if err != errExecutionReverted {
			gas_left = 0
		}
	}
```

it seems to be ok